File: day-1/solution-part2.py
def count_zeros_during_rotation(start, direction, distance):
    """Count how many times the dial points at 0 during a rotation"""
    zero_count = 0
    
    if direction == 'L':
        # Moving left (decreasing)
        for i in range(1, distance + 1):
            position = (start - i) % 100
            if position == 0:
                zero_count += 1
    else:  # direction == 'R'
        # Moving right (increasing)
        for i in range(1, distance + 1):
            position = (start + i) % 100
            if position == 0:
                zero_count += 1
    
    return zero_count

def solve_dial_puzzle_part2(rotations):
    position = 50
    total_zero_count = 0
    
    for rotation in rotations:
        direction = rotation[0]
        distance = int(rotation[1:])
        
        # Count zeros during this rotation
        zeros_during = count_zeros_during_rotation(position, direction, distance)
        total_zero_count += zeros_during
        
        # Update position
        if direction == 'L':
            position = (position - distance) % 100
        else:  # direction == 'R'
            position = (position + distance) % 100
    
    return total_zero_count

# Read your puzzle input
with open('solution.csv', 'r') as f:
    rotations = [line.strip() for line in f.readlines()]

password = solve_dial_puzzle_part2(rotations)
print(f"The password is: {password}")

File: day-1/solution-part2.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;

/// Counts num of times dial points at 0 during rotation
fn count_zeros_during_rotation(start: i32, direction: char, distance: i32) -> i32 {
    let mut zero_count = 0;

    match direction {
        'L' => {
            for i in 1..=distance {
                let position = (start - i).rem_euclid(100);
                if position == 0 {
                    zero_count += 1;
                }
            }
        }
        'R' => {
            for i in 1..=distance {
                let position = (start + i).rem_euclid(100);
                if position == 0 {
                    zero_count += 1;
                }
            }
        }
        _ => panic!("Invalid direction"),
    }

    zero_count
}

/// Solves puzzle
fn solve_dial_puzzle_part2(rotations: Vec<String>) -> i32 {
    let mut position: i32 = 50;
    let mut total_zero_count: i32 = 0;

    for rotation in rotations {
        let direction = rotation.chars().next().unwrap();
        let distance: i32 = rotation[1..].parse().unwrap();

        // Counts zeros during rotation
        let zeros_during = count_zeros_during_rotation(position, direction, distance);
        total_zero_count += zeros_during;

        // Updates position
        position = match direction {
            'L' => (position - distance).rem_euclid(100),
            'R' => (position + distance).rem_euclid(100),
            _ => panic!("Invalid direction"),
        };
    }

    total_zero_count
}

/// Reads file lines into Vec<String>
fn read_lines<P>(filename: P) -> io::Result<Vec<String>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    let buf = io::BufReader::new(file);
    Ok(buf.lines().filter_map(Result::ok).collect())
}

fn main() {
    let rotations = read_lines("solution.csv").expect("Failed to read solution.csv");

    let password = solve_dial_puzzle_part2(rotations);
    println!("The password is: {}", password);
}
